### ISSUE TITLE
v2: Discussion: `strip` directive to replace `strip_prefix` and `strip_suffix`

### ISSUE THREAD

**Issue by whitestrake** [caddyserver/caddy#3140]
`caddy version`: v2.0.0-beta9.0.20200311221200-cfe85a9fe625 (built to latest commit as of this issue, no modules)

This Caddyfile works:

```
http:// {
  route /foo* {
    strip_prefix foo
  }
}
```

This Caddyfile does not:

```
http:// {
  route /foo* {
    strip_prefix /foo
  }
}
```

The error given is:

`2020/03/13 00:23:51.326	INFO	using adjacent Caddyfile
run: adapting config using caddyfile: parsing caddyfile tokens for 'route': Caddyfile:3 - Error during parsing: parsing caddyfile tokens for 'strip_prefix': Caddyfile:3 - Error during parsing: Wrong argument count or unexpected line ending after 'strip_prefix'`

**Comment by mholt** [comment]
Ah, just a hunch real quick, I bet that's because path matchers have to start with a forward slash... so it thinks the first argument is a path matcher, and not the prefix that should be stripped. https://caddyserver.com/docs/caddyfile/concepts#path-matcher -- I think even the examples demonstrate this: https://caddyserver.com/docs/caddyfile/directives/strip_prefix#examples

Let me know if there's any other questions!

**Comment by whitestrake** [comment]
That explains the cause, but do you not consider this a bug?

Providing one valid prefix works, a different valid prefix to strip does not work. The documentation specifically outlines it as valid syntax, and separately explicitly acknowledges a leading slash as the norm: "This value may omit the leading forward slash `/` and it will be assumed."

https://caddyserver.com/docs/caddyfile/directives/strip_prefix

At bare minimum, the Caddyfile parser error is misleading, if not outright incorrect. If we want to keep this unexpected and unintuitive behaviour, should we at least update the documentation to advise as such?

I definitely don't feel that the example sufficiently demonstrates that a single argument to `strip_prefix` that contains a leading slash will generate this error. As far as I know, it's just showing off that you _may_ omit the slash, as mentioned above.

**Comment by mholt** [comment]
To clarify:

One prefix is a prefix to match (just like any other directive), another prefix is the prefix to strip. It's a bit confusing in this case since both arguments are about prefixes. :)

The strip_prefix directive does not _have_ to accept a matcher. We could remove that argument entirely, so that the prefix to match is always the same as the prefix to match. However, then it becomes impossible to filter requests by anything other than path, without using another directive like `route` or `handle` first, i.e. `strip_prefix @post /foo` becomes impossible...

So, matcher token or no matcher token? Those are our options at this point.

I think the current implementation is correct, if using matchers.

**Comment by francislavoie** [comment]
I think requiring `strip_prefix * /foo` is fine, but I think the error messages need blanket improvements across the board to make it clearer to users.

**Comment by whitestrake** [comment]
So after discussing further with @mholt I've come to understand that resolving this to work exactly how the current syntax specifies will come up is technically difficult. This is because the Caddyfile parser attempts to locate a matcher as the first argument before it takes the remaining arguments and passes them to the directive. This is uniform across the entire Caddyfile. So, altering this to handle this specific case is not really feasible.

Rather than the obvious options remaining:
- Document this edge case and leave it alone
- Remove the optionality of the matcher (i.e. always require one)
- Remove the matcher entirely (i.e. never accept a matcher)

Which all have their own downsides... I proposed to Matt that we flip the script on the `strip_x` directives and pick a new syntax that can never encounter this problem:

`strip [<matcher>] <prefix|suffix> <string>`

Matt immediately enumerated a few possible examples:

`strip @post prefix /foo`
`strip suffix /my-suffix`

The question was also raised as to whether `strip_uri` would be more appropriate. While it's good to specify URI here, just for clarity's sake, and it also becomes consistent with the `replace_uri` directive, I'm personally rooting for just `strip` because it's simpler, and it reads more like intuitive English when you're reading one of these directives in a Caddyfile.

Thoughts?

**Comment by francislavoie** [comment]
Oooh, I like that idea of a `strip` directive. Good idea!!

**Comment by mholt** [comment]
@francislavoie The only problem with using a generic name like `strip` is that no other directive -- even if it deals with something other than the path -- can be named that. For example, maybe a directive that strips content from the response body? That's why `replace_uri` directive is called what it is...

So I'm leaning toward `strip_uri` or actually probably `strip_path` since it only deals with the path portion of the URI.

**Comment by francislavoie** [comment]
How about `strip [<matcher>] <path_prefix|path_suffix|uri_whatever> <string>`

**Comment by whitestrake** [comment]
While nothing _immediately_ comes to mind as being something else someone would want to _strip_ in the context of request handling, I suppose being specific with the directive name is the responsible course of action here in case something does pop up.

It just irks me because `strip` is quite a short, neat word and I feel like having things be short, neat, and intuitive/readable is one of the main values of the Caddyfile.

I'd take something like that first example, @francislavoie - although just `<path_prefix|path_suffix>` would probably suffice because altering arbitrarily-placed content comes under the domain of `replace_uri`, which in fact covers all of these cases, just that stripping URI prefixes and suffixes are such a common task they deserved their own shorthand.

Definitely don't love borrowing regex anchors.

**Comment by francislavoie** [comment]
> although just <path_prefix|path_suffix> would probably suffice because altering arbitrarily-placed content comes under the domain of replace_uri

I agree, was just trying to address @mholt's point specifically about `strip` no longer being able to be used for other purposes

> Definitely don't love borrowing regex anchors.

👍

**Comment by francislavoie** [comment]
I might take a shot at implementing this over the weekend if someone else doesn't do it first 😄

**Comment by mholt** [comment]
@francislavoie @whitestrake 

How about a single `uri` directive that unifies:

- `strip_prefix`
- `strip_suffix`
- `uri_replace`

Proposed syntax:

```
uri [<matcher>] strip_prefix|strip_suffix|replace <needle> [<replacement>]
```

Examples:

```
# these first two are equivalent by chance,
# simply because a path that doesn't have
# that prefix won't be stripped of it!
uri /foo/* strip_prefix /foo
uri strip_prefix /foo

# only strips suffix if ends with /bar
uri strip_suffix /bar 

# only does replacement if matches @post and has "foo"
uri @post replace foo bar
```

Sound good? If so, I can get on that today.

**Comment by francislavoie** [comment]
Yep, sounds good to me. Sorry I didn't get around to this yet, wanted to try implementing but if you have the time, go for it! 😄

**Comment by mholt** [comment]
@francislavoie @whitestrake I've implemented these changes in #3157! Please take a look, and I think I can release beta 18 today.
